go/arrow/array/diff.go

```diff
@@ -0,0 +1,248 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package array
+
+import (
+	"fmt"
+
+	"github.com/apache/arrow/go/v12/arrow"
+)
+
+// Diff compares two arrays, returning an edit script which expresses the difference
+// between them. The edit script can be applied to the base array to produce the target.
+//
+// An edit script represents the difference between two arrays.
+// Each element of "insert" determines whether an element was inserted into (true)
+// or deleted from (false) base. Each insertion or deletion is followed by a run of
+// elements which are unchanged from base to target; the length of this run is stored
+// in RunLength. (Note that the edit script begins and ends with a run of shared
+// elements but both fields of the struct must have the same length. To accommodate this
+// the first element of "insert" should be ignored.)
+//
+// For example for base "hlloo" and target "hello", the edit script would be
+// [
+//
+//	{"insert": false, "run_length": 1}, // leading run of length 1 ("h")
+//	{"insert": true, "run_length": 3}, // insert("e") then a run of length 3 ("llo")
+//	{"insert": false, "run_length": 0} // delete("o") then an empty run
+//
+// ]
+// base: baseline for comparison
+// target: an array of identical type to base whose elements differ from base's
+func Diff(base, target arrow.Array) (inserts []bool, runLengths []int64, err error) {
+	if !arrow.TypeEqual(base.DataType(), target.DataType()) {
+		return nil, nil, fmt.Errorf("%w: only taking the diff of like-typed arrays is supported", arrow.ErrNotImplemented)
+	}
+	switch base.DataType().ID() {
+	case arrow.EXTENSION:
+		return Diff(base.(ExtensionArray).Storage(), target.(ExtensionArray).Storage())
+	case arrow.DICTIONARY:
+		return nil, nil, fmt.Errorf("%w: diffing arrays of type %s is not implemented", arrow.ErrNotImplemented, base.DataType().String())
+	case arrow.RUN_END_ENCODED:
+		return nil, nil, fmt.Errorf("%w: diffing arrays of type %s is not implemented", arrow.ErrNotImplemented, base.DataType().String())
+	}
```

Comment:
You don't need the `.String()` in there, because you're using `%s` in the format string, it'll automatically know to call the `String()` method.

Reply:
Good point, fixed!